src/python/pants/core/util_rules/partitions.py

```diff
@@ -49,8 +49,15 @@ def description(self) -> str:
 PartitionElementT = TypeVar("PartitionElementT")
 
 
+@dataclass(frozen=True)
+@runtime_ignore_subscripts
+class Partition(Generic[PartitionKeyT, PartitionElementT]):
+    key: PartitionKeyT | None
```

Comment:
This does nothing because `PartitionKeyT` already can be `None` :smile:

Reply:
That doesn't seem to be the case - if I remove the `| None` here then `mypy` complains about a mismatch between `Optional[PartitionKeyT]` and `PartitionKeyT`

Reply:
The `bound` on the `TypeVar` is my guess for what's preventing the use of `None` as an instance of the generic type 🤔